### PR TITLE
rbac for messages and related

### PR DESCRIPTION
**Pull request by SherinV**
Signed-off-by: Sherin Varughese <shvarugh@redhat.com>

**Related Issue:**  https://github.com/stolostron/backlog/issues/21298

### Description of changes
Split some code into https://github.com/stolostron/search-v2-api/pull/90 and https://github.com/stolostron/search-v2-api/pull/89

### CODE REVIEW

File: pkg/rbac/sharedData.go [hunk 1]
```diff
@@ -16,12 +16,16 @@ import (
 // Cache data shared across all users.
 type SharedData struct {
 	// These are the data fields.
-	csResources     []Resource // Cluster-scoped resources (ie. Node, ManagedCluster)
-	csResourcesMap  map[Resource]struct{}
-	namespaces      []string
-	managedClusters []string
-
+	// csResources      []Resource // Cluster-scoped resources (ie. Node, ManagedCluster)
```
Comment (SherinV) [inline pkg/rbac/sharedData.go:19]: Will remove commented out csResources array once change is finalized.

File: pkg/rbac/userData.go [hunk 2]
```diff
@@ -86,28 +88,32 @@ func (cache *Cache) GetUserData(ctx context.Context,
...
 		klog.V(5).Info("Using user data from cache.")
+
 		return cachedUserData, nil
 	} else {
 		// User not in cache , Initialize and assign to the UID
 		user = &UserDataCache{}
-		cache.users[uid] = user
+		if cache.users != nil {
+			cache.users[uid] = user
+		}
```
Comment (jlpadilla) [inline pkg/rbac/userData.go:98]: Need to implement else. Initialize cache.users and add the user.

Comment (jlpadilla) [reply pkg/rbac/userData.go:98]: Nevermind, this is correct. I was multitasking. :-(

File: pkg/rbac/sharedData.go [hunk 3]
```diff
@@ -258,21 +255,18 @@ func (cache *Cache) GetDisabledClusters(ctx context.Context) (*map[string]struct
...
+		if _, userHasAccess := userClusters[disabledCluster]; userHasAccess { //user has access
+			klog.V(7).Info("user has access to search addon disabled cluster: ", disabledCluster)
+			userAccessDisabledClusters[disabledCluster] = struct{}{}
 		}
 	}
 	return userAccessDisabledClusters
 }
 
-func (cache *Cache) setDisabledClusters(disabledClusters map[string]struct{}, err error) {
+func (cache *Cache) SetDisabledClusters(disabledClusters map[string]struct{}, err error) {
```
Comment (SherinV) [inline pkg/rbac/sharedData.go:269]: Testing something. Will unexport this function once done.

File: pkg/rbac/messagesDev.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
+// Copyright Contributors to the Open Cluster Management project
+
```
Comment (SherinV) [inline pkg/rbac/messagesDev.go:2]: @jlpadilla What tags should be here? Is it same as https://github.com/stolostron/search-indexer/blob/main/pkg/config/config_development.go#L3-L7

Comment (jlpadilla) [reply pkg/rbac/messagesDev.go:2]: Maybe `test`?  You'll also need to update the Makefile to include this tag like here.
https://github.com/stolostron/search-indexer/blob/main/Makefile#L13-L14

Comment (jlpadilla) [reply pkg/rbac/messagesDev.go:2]: Also, try naming the file `*_test.go`.  That has a similar effect, but not sure if it works across packages.

Comment (SherinV) [reply pkg/rbac/messagesDev.go:7]: Done adding tags.

File: pkg/rbac/userData.go [hunk 5]
```diff
@@ -249,16 +257,26 @@ func (user *UserDataCache) getNamespacedResources(cache *Cache, ctx context.Cont
...
-							if !cache.shared.isClusterScoped(res, api) { //Add the resource if it is not cluster scoped
+							//Add the resource if it is not cluster scoped
+							//fail-safe mechanism to avoid whitelist - TODO: incorporate whitelist
+							if !cache.shared.isClusterScoped(res, api) && (len(rules.ResourceNames) == 0 ||
+								(len(rules.ResourceNames) > 0 && rules.ResourceNames[0] == "*")) {
 								user.userData.NsResources[ns] = append(user.userData.NsResources[ns],
 									Resource{Apigroup: api, Kind: res})
+							} else if cache.shared.isClusterScoped(res, api) {
+								klog.V(5).Info("Got clusterscoped resource", api, "/",
+									"res from SelfSubjectRulesReviews. Excluding it from ns scoped resoures.")
```
Comment (Anxhela21) [inline pkg/rbac/userData.go:275]: do you mean to print the value for res in the log here?

File: pkg/rbac/userData_test.go [hunk 6]
```diff
@@ -568,3 +590,50 @@ func Test_managedCluster_GetUserData(t *testing.T) {
...
+		nsrUpdatedAt: time.Now(),
+	}
+	ctx := context.Background()
+	ctx = context.WithValue(ctx, ContextAuthTokenKey, "123456")
+
+	// result, err := mock_cache.GetUserDataCache(ctx, nil)
+
+	result, err := mock_cache.GetUserData(ctx)
+	fmt.Println("res: ", result)
+	fmt.Println("err: ", err)
```
Comment (Anxhela21) [inline pkg/rbac/userData_test.go:627]: delete these and remove comment above.

File: pkg/rbac/sharedData_test.go [hunk 7]
```diff
@@ -181,11 +185,52 @@ func Test_getResources_expiredCache(t *testing.T) {
...
+	}
+}
+
+func Test_GetandSetDisabledClusters(t *testing.T) {
+	_, mock_cache := mockResourcesListCache(t)
+	mock_cache.shared.dcUpdatedAt = time.Now()
+
+	dClusters := make(map[string]struct{})
+	dClusters["managed1"] = struct{}{}
+	dClusters["managed2"] = struct{}{}
```
Comment (jlpadilla) [inline pkg/rbac/sharedData_test.go:211]:
```suggestion
	dClusters := map[string]struct{} {"managed1": struct{}{}, "managed2": struct{}{}}
```

File: pkg/rbac/userData.go [hunk 8]
```diff
@@ -170,14 +173,15 @@ func (user *UserDataCache) getClusterScopedResources(cache *Cache, ctx context.C
...
 			user.userData.CsResources = append(user.userData.CsResources,
 				Resource{Apigroup: res.Apigroup, Kind: res.Kind})
 		}
 	}
-	uid, _ := cache.GetUserUID(ctx)
-	klog.V(7).Infof("User %s has access to these cluster scoped res: %+v \n", uid,
+	uid, userInfo := cache.GetUserUID(ctx)
+	klog.V(7).Infof("User %s with uid: %s has access to these cluster scoped res: %+v \n", userInfo.Username, uid,
+
 		user.userData.CsResources)
```
Comment (jlpadilla) [inline pkg/rbac/userData.go:185]: Delete extra line
```suggestion
	klog.V(7).Infof("User %s with uid: %s has access to these cluster scoped res: %+v \n", userInfo.Username, uid,
		user.userData.CsResources)
```

File: pkg/rbac/userData.go [hunk 9]
```diff
@@ -252,16 +257,26 @@ func (user *UserDataCache) getNamespacedResources(cache *Cache, ctx context.Cont
...
+			klog.V(9).Infof("SelfSubjectRulesReviews Kube API result for ns:%s : %v\n", ns, prettyPrint(result.Status))
 		}
 		for _, rules := range result.Status.ResourceRules {
 			for _, verb := range rules.Verbs {
 				if verb == "list" || verb == "*" { //TODO: resourceName == "*" && verb == "*" then exit loop
 					for _, res := range rules.Resources {
 						for _, api := range rules.APIGroups {
-							if !cache.shared.isClusterScoped(res, api) { //Add the resource if it is not cluster scoped
+							//Add the resource if it is not cluster scoped
+							//fail-safe mechanism to avoid whitelist - TODO: incorporate whitelist
```
Comment (jlpadilla) [inline pkg/rbac/userData.go:268]: FYI, I added this to our umbrella bug issue.

File: pkg/rbac/userData_test.go [hunk 10]
```diff
@@ -578,3 +583,48 @@ func Test_managedCluster_GetUserData(t *testing.T) {
...
+	managedClusters["some-managed-cluster"] = struct{}{}
+	managedClusters["some-other-managed-cluster"] = struct{}{}
+
+	mock_cache.users["unique-user-id"] = &UserDataCache{
+		userData:          UserData{CsResources: allowedres, ManagedClusters: managedClusters, NsResources: nsRes},
+		clustersUpdatedAt: time.Now(),
+		// Using current time , GetUserData should have the same values as cache
+		csrUpdatedAt: time.Now(),
+		nsrUpdatedAt: time.Now(),
+	}
```
Comment (jlpadilla) [inline pkg/rbac/userData_test.go:610]: I find it easier to understand when the mock data is defined inline. Thoughts?

```suggestion
	mock_cache.users["unique-user-id"] = &UserDataCache{
		userData: UserData{
			CsResources:     []Resource{{Apigroup: "storage.k8s.io", Kind: "nodes"}},
			ManagedClusters: map[string]struct{}{"some-managed-cluster": {}, "some-other-managed-cluster": {}},
			NsResources:     map[string][]Resource{"ns1": []Resource{{Apigroup: "", Kind: "pods"}}},
		},
		// Using current time , GetUserData should have the same values as cache
		clustersUpdatedAt: time.Now(),
		csrUpdatedAt:      time.Now(),
		nsrUpdatedAt:      time.Now(),
	}
```

Comment (jlpadilla) [reply pkg/rbac/userData_test.go:610]: The existing tests follow this pattern so I'm also good with consistency.

File: pkg/rbac/userData_test.go [hunk 11]
```diff
@@ -578,3 +583,48 @@ func Test_managedCluster_GetUserData(t *testing.T) {
...
+	ctx := context.Background()
+	ctx = context.WithValue(ctx, ContextAuthTokenKey, "123456")
```
Comment (jlpadilla) [inline pkg/rbac/userData_test.go:612]: Maybe simpler.
```suggestion
	ctx := context.WithValue(context.Background(), ContextAuthTokenKey, "123456")
```

File: pkg/resolver/related.go [hunk 12]
```diff
@@ -134,9 +134,14 @@ func (s *SearchResult) buildRelationsQuery() {
 	// add a union to the relation query to get all resources in the clusters
 	clusterSelectTerm := s.selectIfClusterUIDPresent()
 	if clusterSelectTerm != nil {
-		relQuery = relQuery.Union(clusterSelectTerm)
+		relQuery = relQuery.Union(clusterSelectTerm).As("related")
 	}
-	sql, params, err := relQuery.ToSQL()
+	relQuery = goqu.From(relQuery.As("related")).Select("related.uid", "related.kind",
+		"related.level") //goqu.L(`"srchAddon".uid`)
```
Comment (jlpadilla) [inline pkg/resolver/related.go:140]: Should this comment be here?

File: pkg/resolver/related_test.go [hunk 13]
```diff
@@ -61,23 +62,24 @@ func Test_SearchResolver_RelationshipsWithCluster(t *testing.T) {
...
 	uid1 := "cluster__local-cluster"
-
+	csRes, nsRes, mc := newUserData()
+	ud := rbac.UserData{CsResources: csRes, NsResources: nsRes, ManagedClusters: mc}
 	resultList = append(resultList, &uid1)
 
 	// take the uids from above as input
 	searchInput := &model.SearchInput{Filters: []*model.SearchFilter{{Property: "kind", Values: resultList}}}
-	ud := rbac.UserData{}
+	// ud := rbac.UserData{}
```
Comment (jlpadilla) [inline pkg/resolver/related_test.go:71]: Delete commented line.

File: pkg/resolver/search_test.go [hunk 14]
```diff
@@ -124,7 +125,7 @@ type TestOperatorItem struct {
 }
 
 func Test_SearchResolver_ItemsWithNumOperator(t *testing.T) {
-	// rbac := buildRbacWhereClause(context.TODO(), &rbac.UserResourceAccess{csres, nsres, mc})
+	// rbac := buildRbacWhereClause(context.WithValue(context.Background(), rbac.ContextAuthTokenKey, "123456"), &rbac.UserResourceAccess{csres, nsres, mc})
```
Comment (jlpadilla) [inline pkg/resolver/search_test.go:128]: Delete commented code.

**Review by jlpadilla — APPROVED** [review]
/hold
/lgtm
Please review remaining comments, all are minor cleanup or style suggestions.

File: pkg/rbac/userData_test.go [hunk 15]
```diff
@@ -578,3 +575,58 @@ func Test_managedCluster_GetUserData(t *testing.T) {
...
+	// managedClusters["some-managed-cluster"] = struct{}{}
+	// managedClusters["some-other-managed-cluster"] = struct{}{}
+
+	// mock_cache.users["unique-user-id"] = &UserDataCache{
+	// 	userData:          UserData{CsResources: allowedres, ManagedClusters: managedClusters, NsResources: nsRes},
+	// 	clustersUpdatedAt: time.Now(),
+	// 	// Using current time , GetUserData should have the same values as cache
+	// 	csrUpdatedAt: time.Now(),
+	// 	nsrUpdatedAt: time.Now(),
+	// }
```
Comment (jlpadilla) [inline pkg/rbac/userData_test.go:602]: Please delete

**Review by jlpadilla — APPROVED** [review]
/lgtm

**Pull request merged**